src/main.cpp

```diff
@@ -2631,6 +2631,26 @@ static bool ActivateBestChainStep(CValidationState &state, CBlockIndex *pindexMo
     const CBlockIndex *pindexOldTip = chainActive.Tip();
     const CBlockIndex *pindexFork = chainActive.FindFork(pindexMostWork);
 
+    auto reorgLength = pindexOldTip->nHeight - pindexFork->nHeight;
+    if (reorgLength > MAX_REORG_LENGTH) {
+        auto msg = strprintf(_(
+            "A block chain reorganization has been detected that would roll back %d blocks! "
+            "This is larger than the maximum of %d blocks, and so the node is shutting down for your safety."
+            ), reorgLength, MAX_REORG_LENGTH) + "\n\n" +
+            _("Fork details") + ":\n" +
+            "- " + strprintf(_("Current tip: %s, height %d, work %s"),
+                pindexOldTip->phashBlock->GetHex(), pindexOldTip->nHeight, pindexOldTip->nChainWork.GetHex()) + "\n" +
```

Comment:
I can't find some of the text you gave in the example 
"To stay on the current chain,.."

Reply:
It was all removed - see https://github.com/zcash/zcash/pull/2463#issuecomment-362870103

Reply:
The current message seems unclear it says "human help" but seems totally unclear as to what the human should do. I also had a hard time understanding from this comment https://github.com/zcash/zcash/issues/713#issuecomment-361397597  what the user should do.
How exactly does deciding what software upgrade to do have to do with choosing a fork?
There might be a > 100 reorg unrelated to any software upgrades

Reply:
You talk in the comment about checkpoints, but are you planning to have a checkpoint every 100 blocks?

Reply:
> You talk in the comment about checkpoints, but are you planning to have a checkpoint every 100 blocks?

You don't need a checkpoint every 100 blocks; just whenever a rollback of at least 100 blocks occurs. That's large enough to likely affect multiple people, so once the chain to use has been decided, devs can push out an update with a checkpoint to that effect, and then users can reindex.

> How exactly does deciding what software upgrade to do have to do with choosing a fork?
There might be a > 100 reorg unrelated to any software upgrades

See above: the upgrades referred to happen as a result of the reorgs. The choice of software upgrade would imply choosing which branch of the fork to take, based on which checkpoint the update includes.

Reply:
(Answering @arielgabizon) Not as a result of, but yes all forks would *require* an upgrade in order to be resolved under this policy.